### ISSUE TITLE
Allow typing in values outside of min/max slider range

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#43]
For sliders, could you provide an option per slider to allow user to type in and accept a value that is outside the range of the slider?

**Comment by RobinSchmidt** [comment]
phew -  that's actually totally against the rules. i need to think about the consequences...

**Comment by elanhickler** [comment]
robin, how about something simple: separate slider range "slidable range" and slider range "actual range"

We can have two ranges for sliders.

**Comment by elanhickler** [comment]
We need multiple ranges anyway, I'm think I want to have a right click menu "extend range" option for some sliders where some power users may want extreme range while others will want "easy mode" ranges.

**Comment by RobinSchmidt** [comment]
hmm...well - i have something like this in func-shaper. there you can adjust min/max values of the a,b,c,d sliders. do you think, such functionality should be enclosed into a class? we could make a slider subclass with min/max textfield widgets

**Comment by RobinSchmidt** [comment]
or maybe not a subclass but one that contains a slider and 2 textfields

**Comment by elanhickler** [comment]
I just thought of something. We had a discussion on changing parameter ranges and ruining presets at GitLab I believe. What if you stored the ranges as part of the patch xml system? Then later you could also introduce the ability to change slider ranges. There's a lot of times I wish I could change parameter ranges for specific situations. If the user needs a larger range for convenience, it's... well... very convenient.

**Comment by RobinSchmidt** [comment]
the main reason is that i don't want to bloat and clutter the preset files. my reasoning is that it's rather unlikely to change a parameter range after a product has been published. i, as a programmer, have the responsibility to think about the appropriate range beforehand. and if that should really be necessary to change it at some point, i can still resort to the preset versioning and explicit conversion. i also don't store parameter values, when they are equal to their default values - also for economy reasons

**Comment by elanhickler** [comment]
for me it's necessary as there's still missing features in the library that prevents me from using the ranges I'd like to use. If the preset could store ranges I can wait comfortably while those features are being added and change the ranges later.

**Comment by RobinSchmidt** [comment]
> No idea how to do version converting of presets.

you retrieve the PatchFormat attribute from the xml, and compare the returned value to the AudioModule's patchFormatIndex. if the patch format in the xml is older, make appropriate changes to the relevant xml-attibutes. look at PitchShifterAudioModule::convertXmlStateIfNecessary for an example (although, i don't really compare to the patchFormatIndex member there - i just loo, if the returned value is equal to 0 and then i already know, what to do). 

well, it's a bit messy. 

i could provide hook-functions like storeParameterToXml/recallParameterFromXml which you could override and then store/recall min/max values there. ...i.e. make your own baseclass for all your AudioModules (subclass of my ModulatableAudioModule class)

but....i think, this will still mess up presets, when they are connected to meta-parameters

**Comment by elanhickler** [comment]
I'm no longer going to connect meta-parameters by default.

**Comment by elanhickler** [comment]
The problem I'm having right now is that I have a frequency knob with range -5000 to +5000. I want bipolar frequency and 0 frequency allowed. Problem is that the usual range of LFOs is between -10 and +10hz, so I'd probably want to do exponential scaling, but you can't have negatives and zeros.

I think this scaling mode could work: https://www.desmos.com/calculator/vfjfcdjout

pow3 scaling.

**Comment by RobinSchmidt** [comment]
i think, it's time to facilitate the use of custom mapping function for parameters. a bipolar version of the exponential mapping is something quite particular and it needs at least one additional parameter to specify its behavior. it could be based on my linToExpWithOffset function (and the corresponding expToLinWithOffset inverse function - we always need two-way mapping). this specifies another "offset" parameter that determines the curve/slope at zero. you could use the regular 20...20000 mapping with an offset of -20 and get roughly the same exp-scaling behavior, but from 0 to 20000. and then that could be mirrored for negative values. however, i think, this is too much to pack into the parameter class. most of the time, it will be just unneeded bloat in the objects

**Comment by RobinSchmidt** [comment]
ok - my Parameter class supports now custom mapping functions. this was a sort of open-heart surgery of one of my most basic framework classes, so some care had to be taken and it will need some good testing. it has a new member function:

`virtual void setMapper(rsParameterMapper* newMapper);`

which you can use to pass an object of a subclass of rsParameterMapper into. your subclass must implement `map` and `unmap` for forward and backward mapping (unmap should be the inverse function of map, for example if map is exp and unmap should be log). however, you don't need to use this. the old ways (of selecting one of the enumerated scalings) still work - but what happens internally is that an appropriate mapper object is then created by the Parameter object).

**Comment by RobinSchmidt** [comment]
so now the mapping is completely flexible and you can have any mapping you want by implementing your own mapper subclass and configuring the parameter with it.

...as for the bidirectional exponential mapping: i have thought about this and i think now, the best (most natural, smooth) function for this would be something based on sinh. for example y = a * sinh(b*x) for some constants a,b chosen appropriately to cover the desired range and have the desired slope at the origin (the lower the value of b, the more linear it becomes). i think, this is the next task: working out the details of the sinh mapping. maybe it could be further flexibilized by using y = a * sinh(b*(x+c)) + d

**Comment by elanhickler** [comment]
no robin! I only understand pictures, you have to make desmos 😸 

https://www.desmos.com/calculator/vpnilxqab0

this feels right to me.

**Comment by elanhickler** [comment]
OH WAIT A MINUTE, bipolar exponential FM might be f***ing amazing. Linear FM stays in tune, exponential FM does not, because lower values have less change in frequency unlike linear where you get the same amount of change in frequency whether you go higher or lower.

Wow, what if we did a crossfade from bipolar expo / bipolar linear mapper for use as an FM frequency controller? That could make for some awesome sound synthesis.

**Comment by RobinSchmidt** [comment]
YES!!! that sounds like some interesting stuff to explore!

**Comment by elanhickler** [comment]
mapper rsParameterMapperSinh doesn't seem to work, slider just goes to 0.

also you gotta fix this bug:

![image](https://user-images.githubusercontent.com/9116575/31754335-c8d8755a-b44b-11e7-8939-34f01b168793.png)

you have your numscalings before CUSTOM which throws an error becaue you check to make sure you're not using a scaling that is more than numscalings.

also, could you make setShape allow for chaning a AND b?

**Comment by RobinSchmidt** [comment]
> mapper rsParameterMapperSinh doesn't seem to work, slider just goes to 0 when moved.

yes, i know - i said that in the other thread. the patch to the slider class was actually simple. should work now.

as for "CUSTOM" being below NUM_SCALINGS: that's actually intentional: the custom scaling is not among the enumerated ones, so to speak. i think, i do this check only when you call setScaling - right? ...and you are not supposed to call that with "CUSTOM". the corresponding member will be set to CUSTOM, if you pass in a mapper object.

btw.: when you use enumerated values, ALWAYS ALWAYS ALWAYS  use the enumeration values. NEVER EVER EVER do things like: myParam->setScaling(2). i may change the order of the enumerations at some point, insert new items, etc. - this will then break your code

**Comment by RobinSchmidt** [comment]
the same goes also for filter types in rosic. i've seen you using such "magic numbers" before

**Comment by RobinSchmidt** [comment]
> could you make setShape allow for chaning a AND b?

nope. a must be a function of b. they are coupled, otherwise, the min/max values come out wrong

**Comment by RobinSchmidt** [comment]
basically, you set b and max, and that fixes a

**Comment by elanhickler** [comment]
how would I get this shape?: https://www.desmos.com/calculator/vpnilxqab0

**Comment by RobinSchmidt** [comment]
this exact function? well, you already know you b-value. it's one in this case. you want to know how to set max? just read off the value of the graph at x=1 - it's around 0.87. i don't know what you are trying to achieve...

**Comment by elanhickler** [comment]
I'm going to use this curve for bipolar frequency and want to make an rsParameterMapper subclass, and need to setup the map and unmap functions.

**Comment by elanhickler** [comment]
https://gitlab.com/Hickler/Soundemote/issues/67

https://www.kvraudio.com/forum/viewtopic.php?p=6656238#p6656238

https://github.com/RobinSchmidt/RS-MET/issues/72

**Comment by RobinSchmidt** [comment]
yes! thank you!

**Comment by elanhickler** [comment]
oops wrong thread?

**Comment by RobinSchmidt** [comment]
nevermind, i'll copy the links into my code-comments anyway. this is intriguing

**Comment by RobinSchmidt** [comment]
> I'm going to use this curve for bipolar frequency and want to make an rsParameterMapper subclass, and need to setup the map and unmap functions.

why do you want to make a subclass and not just use the rsParameterMapperSinh class? just set your min/max frequency value, like to -20000 to + 20000 or whatever and then set up your b (shape) value in a way which nicely balances the precision around zero with the precision at +-20k. this is exactly what this is made for. like 

`freqParam->setMapper(new rsParameterMapperSinh(-20000, +20000, 2.0));`

the last parameter, here 2.0, is "b" and it sets the shape, i.e. determines the trade-off between low-freq and high-freq precision of the slider

**Comment by elanhickler** [comment]
sooooo, I set the min max of the mapper, but that is NOT the min/max of the slider? It is simply the min/max of controlling curvature or someting?

**Comment by RobinSchmidt** [comment]
the min/max of the mapper is the min/max of the parameter is the min/max of the slider. you want a different range for the slider than the parameter has?

**Comment by elanhickler** [comment]
no, i just don't know, given your sinh mapper how to achieve this curve: https://www.desmos.com/calculator/vpnilxqab0

now, it's not like I MUST have this exact curve, but I thought it would be easy to try. No? Anyway, ill use your sinh mapper as is for now.

**Comment by RobinSchmidt** [comment]
i think, there's no much sense in wanting this exact curve. our exponential mappings for frequencies say, 20...20000 are also scaled (input and output wise) so as to hit the correct value 20 and 20000 at 0 and 1. more meaningful might be to solve for a b-value, such that a middle frequency...say 1000Hz is at half of (the right part of) the slider range. this is the only thing, one could do, when the maximum is fixed

**Comment by RobinSchmidt** [comment]
...or wait...you are actually shifting your exp-function to the left...i think, this is different from what our exp-scaling does. it's an exp with shifted input....hmmm...

**Comment by elanhickler** [comment]
don't worry about it, I have no idea what I'm doing. 😄

**Comment by RobinSchmidt** [comment]
make sure that you choose your b-value wisely. later changing it will break the mapping to meta parameters (i.e. may break patches with connected metas and daw automations)

**Comment by elanhickler** [comment]
Yes, I found a good way to think of it, just ask what range is most important.

I want -5000 to +5000, but the important values are perhaps -10 to +10, so I choose a B value where -10 to +10 looks about linear.

**Comment by elanhickler** [comment]
still hitting this:

![image](https://user-images.githubusercontent.com/9116575/31800305-df22656a-b4f4-11e7-84bc-b031941a8b2a.png)

**Comment by elanhickler** [comment]
Is it possible to create a parameter mapper that skips 0 and is integer numbers? So I need a slider that has a sequence like this:

-5 -4 -3 -2 -1 +1 +2 +3 +4 +5

I can try to attempt it once the sinh works.

**Comment by RobinSchmidt** [comment]
> still hitting this

i need to know, which class/function this is. you cut the top off too early ;-)

**Comment by RobinSchmidt** [comment]
here's how you can see the actual mapping function in term of max (m) and b:
https://www.desmos.com/calculator/vpnilxqab0
i chose a range -20000 to +20000 as would be appropriate for frequencies in the audible range

**Comment by RobinSchmidt** [comment]
ok, i'm allowing to set Parameter::CUSTOM from client code now (have dragged the value before NUM_SCALINGS)

**Comment by elanhickler** [comment]
jura_Parameter.cpp line 195

myparams.cpp line 113

I thought setting to custom was what you had to do? not sure how else it'd work.

**Comment by RobinSchmidt** [comment]
as said above, actually the idea was that the user/client doesn't manually sets this option but it is only set internally as soon as you pass a mapper. but i didn't take into account that the constructor also requires this setting....you could actually have passe any value, it would later be updated in the object to "custom" when you pass the mapper anyway

**Comment by elanhickler** [comment]
is it possible make a parameter mapper that skips 0? -5 -4 -3 -2 -1 +1 +2 +3 +4 +5

here's the linear equations:
```
double   map(double x) const override { return min + (max-min) * x; }
double unmap(double y) const override { return (y-min) / (max-min); }
```

if I copy that and add this to a new mapper class, would it work?:
```
x <= 0 ? x = x-1 : x
y >= 0 ? y = y+1 : y
```

**Comment by RobinSchmidt** [comment]
hmm..you need rounding...but then, how would the inverse function unmap look like? such a function is actually not invertible. we need inversion, because when client code calls setValue and a Meta is connected, we have to able to figure out the meta value (which is always in 0..1).

...hmm...don't know, maybe we could somehow avoid the requirement of two-way mapping. ...or you just map the rounded value back

**Comment by elanhickler** [comment]
still not getting a bipolar exponential feel to the sliders, seems to be bipolar linear.

**Comment by RobinSchmidt** [comment]
what's your b-value?

**Comment by elanhickler** [comment]
i've tried .01, .0075, .3, seems to make no difference

**Comment by RobinSchmidt** [comment]
look at this: such low b leads to almost linear scaling:
https://www.desmos.com/calculator/pamgkqwry9

**Comment by elanhickler** [comment]
uhhh... but ranges im dealing with could be -8 to 8, -5000 to 5000, now I'm confused as to how to use desmos to find a good b value. I thought I knew, now I have no idea

**Comment by RobinSchmidt** [comment]
this actually doesn't matter. the shape stays the same, just the whole graphs gets scaled down by another "a" factor then

**Comment by RobinSchmidt** [comment]
this is what i meant when i said: a is fixed by the max-value, as soon as b (shape) is chosen

**Comment by elanhickler** [comment]
ok now i get it!

**Comment by RobinSchmidt** [comment]
as said, i would adjust b in a way such that the graph passes though x = 0.5, y = "some middle value"

**Comment by RobinSchmidt** [comment]
...maybe i could come up with an equation to automatically calculate a b from such a requirement....but i think, it involves solving an implicit equation

**Comment by RobinSchmidt** [comment]
...could certainly be done by newton iteration, though

**Comment by elanhickler** [comment]
The mapper stuff is awesome! Now for the final coup-de-grace.

I need a bipolar log, so the opposite curvature of sinh.. oh jeez, I probably don't have the math knowledge to build a map and unmap function for that, might have to ask you to do it.

**Comment by RobinSchmidt** [comment]
what about tanh?

**Comment by RobinSchmidt** [comment]
for what kind of parameter is this?

**Comment by elanhickler** [comment]
triangle sharpness. As you get toward -1 or +1 sharpness (essentially sawtooth shape) the harmonics increase exponentially, so we need to undo that exponenialness by doing a log curve.

**Comment by RobinSchmidt** [comment]
what was your rational shape formula - do you know this off the cuff? ...the one with a and x or x^2 in numerator and denominator. this came up before several times but i can't memorize the whereabouts. couldn't that be suitable mapping too?

**Comment by elanhickler** [comment]
yes i was just thinking about that, i think we are on the same wavelength of thinking right now.... going to make a desmo for it, one sec.

**Comment by elanhickler** [comment]
https://www.desmos.com/calculator/xaklfkriac

**Comment by RobinSchmidt** [comment]
yes. i guess, it makes sense to provide a mapper using these curves, too. maybe in two versions: unipolar and bipolar

**Comment by RobinSchmidt** [comment]
> Edit: Actually C is the opposite, you need bipolar expo, but A parameter could use bipolar log.
> see: https://www.desmos.com/calculator/7h9mknbv3q

c bipolar xep is what i was saying, iirc. a being sigmoid (it's not a log)...don't know. set b to 1.57 and the linear scaling for a feels just right, imho. it becomes more complex when parameter interact

**Comment by elanhickler** [comment]
> set b to 1.57 and the linear scaling for a feels just right
setting b to 1.57 doesn't change anything, you're still going from no sharpness to infinite sharpness, you should approach that infinity slower.

**Comment by elanhickler** [comment]
see gif: https://puu.sh/y321n/e883d67973.gif

near 1 sharpness, you're not able to fine control the harmonic content, while near 0 you have a lot more control than needed.

**Comment by RobinSchmidt** [comment]
aha! good to know. so it needs some fine tuning of the parameter mapping to make it feel more "musical"

**Comment by elanhickler** [comment]
Hey robin, if you could implement a bipolar-log style mapper today that would be nice because I don't have to break presets for adding it later. You can charge me time for this. I am just about ready to start making presets and audio demos.

**Comment by RobinSchmidt** [comment]
ok - i made a tanh based sigmoid mapper. that was actually so easy, i did it in 5 minutes. it was just a matter of copy/paste the code of the sinh mapper and replace sinh/asinh by tanh/atanh: 

https://www.desmos.com/calculator/ppx5yhza2r

but i didn't test it yet - but i can't imagine anything going wrong

**Comment by RobinSchmidt** [comment]
...but now i wonder, if i could somehow factor out the common code. code duplication sucks

**Comment by RobinSchmidt** [comment]
i still think, your rational function should perhaps be available too ...i think, the exact behavior/trade-off/precision-curve will be very different from tanh, even though looks sort of similar

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/31854172-05df98f2-b64a-11e7-968a-eb221c84fe40.png)

I'm starting to create some "controller profiles" so we can set all this crap all at once,

modulation ranges, modulation limits, slider scalings, etc. just do away with all the confusing "per slider settings" and just describe what the intention is.

**Comment by elanhickler** [comment]
Oh I just realized before we even started working together I was using rational function scaling for filter frequencies, so I could go from 0 to 20,000hz instead of 1.e-6 to 20,000, I found that scheme to be very nice for my chaos generator. I might try to implement the unipolar rational function myself, unless you get to it first.

**Comment by RobinSchmidt** [comment]
ok, do it. you'll probably be faster. i'm in weekend mode now. soon off to a goa party. i think, for frequencies, the exponential scaling is just right. it translates equal distances on the slider to equal multiplication factors in value. i.e. the slider distance between 100 and 200 Hz is the same as between 1000 and 2000 or 10000 and 20000. this is parallels how humans perceive frequency (and a lot of other things):

https://en.wikipedia.org/wiki/Weber%E2%80%93Fechner_law

however, we could perhaps be even more fancy and do some scaling based on that:

https://en.wikipedia.org/wiki/Mel_scale

**Comment by RobinSchmidt** [comment]
slider/parameter profiles sounds like a good idea. ...like just calling:

setParameterProfile(myParameter, FILTER_FREQ)

...which then sets min/max default values, etc. gets rid of a lot of duplicated code

**Comment by RobinSchmidt** [comment]
you are applying the DRY principle! :-D

**Comment by elanhickler** [comment]
before: https://puu.sh/y3V1F/75679b3b09.gif
after: https://puu.sh/y3V1F/75679b3b09.gif

**Comment by elanhickler** [comment]
ok, I tried to implement rational function, but I don't understand what value ranges are being sent to map/unmap function and I can't get a breakpoint on it.
